lib/Sema/SemaBounds.cpp

```diff
@@ -3285,6 +3288,318 @@ namespace {
+    }
+
+    // Compute Gen Sets
+    for (auto B : Blocks) {
+      if (const Stmt *Term = B->Block->getTerminator()) {
+        if(const IfStmt *IS = dyn_cast<IfStmt>(Term)) {
+          ComparisonSet Comparisons;
+          ExtractComparisons(B->Block->getTerminatorCondition(), Comparisons);
+          B->GenThen.insert(Comparisons.begin(), Comparisons.end());
+
+          ComparisonSet NegatedComparisons;
+          Negate(Comparisons, NegatedComparisons);
+          B->GenElse.insert(NegatedComparisons.begin(), NegatedComparisons.end());
+        }
+      }
+      AllComparisons.insert(B->GenThen.begin(), B->GenThen.end());
+      AllComparisons.insert(B->GenElse.begin(), B->GenElse.end());
+    }
+
+    // Compute Kill Sets
+    for (auto B : Blocks) {
+      llvm::SmallPtrSet<const VarDecl *, 16> DefinedVars;
+      for (CFGElement Elem : *(B->Block))
+        if (Elem.getKind() == CFGElement::Statement)
+          CollectDefinedVars(Elem.castAs<CFGStmt>().getStmt(), DefinedVars);
+
+      for (auto E : AllComparisons)
+        for (auto V : DefinedVars)
+          if (ContainsVariable(E, V))
+            B->Kill.insert(E);
+    }
+
+    // Iterative Worklist Algorithm
+    while (!WorkList.empty()) {
+      ElevatedCFGBlock *CurrentBlock = WorkList.front();
+      WorkList.pop();
+
+      // Update In set
+      ComparisonSet IntermediateIntersecions;
+      bool FirstIteration = true;
+      for (CFGBlock::const_pred_iterator I = CurrentBlock->Block->pred_begin(), E = CurrentBlock->Block->pred_end(); I != E; ++I) {
+        if (!*I)
+          continue;
+        if (*((*I)->succ_begin()) == CurrentBlock->Block) {
+          if (FirstIteration) {
+            IntermediateIntersecions = GetByCFGBlock(*I)->OutThen;
+            FirstIteration = false;
+          } else
+            IntermediateIntersecions = Intersect(IntermediateIntersecions, CurrentBlock->OutThen);
+        } else {
+          if (FirstIteration) {
+            IntermediateIntersecions = GetByCFGBlock(*I)->OutElse;
+            FirstIteration = false;
+          } else
+            IntermediateIntersecions = Intersect(IntermediateIntersecions, CurrentBlock->OutElse);
+        }
+      }
+      CurrentBlock->In = IntermediateIntersecions;
+
+      // Update Out Set
+      ComparisonSet OldOutThen = CurrentBlock->OutThen, OldOutElse = CurrentBlock->OutElse;
+      CurrentBlock->OutThen = Difference(Union(CurrentBlock->In, CurrentBlock->GenThen), CurrentBlock->Kill);
+      CurrentBlock->OutElse = Difference(Union(CurrentBlock->In, CurrentBlock->GenElse), CurrentBlock->Kill);
+
+      // Recompute the Affected Blocks
+      if (Differ(OldOutThen, CurrentBlock->OutThen) || Differ(OldOutElse, CurrentBlock->OutElse))
+        for (CFGBlock::const_succ_iterator I = CurrentBlock->Block->succ_begin(), E = CurrentBlock->Block->succ_end(); I != E; ++I)
+          WorkList.push(GetByCFGBlock(*I));
+    }
+
+  //if(DEBUG_DATAFLOW)
```

Comment:
This needs to be moved out to a separate routine.  The dumps should be controlled by a compiler flag.

Reply:
Done.